File: NET/CDOLWTSD/CDODataController.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Threading.Tasks;
using System.Linq;
using DXMPP;

namespace CDOLWTSD
{
    public class CDODataController : IDisposable
    {
        CDO.CDOController SessionController;
        LWTSD.DataController LowlevelDataController;
        Connection Uplink;
        JID Orchestrator;
        string TargetEntityID;

        public ConcurrentDictionary<LWTSD.ResourcePath, CDO.ResourceAccess> RequestedResources =
            new ConcurrentDictionary<LWTSD.ResourcePath, CDO.ResourceAccess>();

        LWTSD.SimplifiedSchema KnownSchema = null;
        public volatile CDO.Session ActiveSession;


        void SessionTerminated(CDO.Session Session)
        {
            Console.WriteLine("Session terminated!");

            if (Session.ID == ActiveSession.ID)
                ActiveSession = null;
        }

        public async Task<bool> RequestNewSession(TimeSpan SessionLength, int TimeOutSeconds = 60)
        {
            Console.WriteLine("Await request session");
            CDO.Session NewSession = await SessionController.RequestSession(
                    TargetEntityID, RequestedResources.Values.ToList(), SessionLength, TimeOutSeconds);
            Console.WriteLine("Done");
            if (NewSession == null)
                throw new AccessViolationException("New session was actively denied");

            JID TargetJID = await SessionController.StartSession(NewSession.ID);
            if (TargetJID == null)
                throw new InvalidProgramException("Session failed to start");

            if (LowlevelDataController != null)
            {
                LowlevelDataController.Dispose();
                LowlevelDataController = null;
            }

            LowlevelDataController = new LWTSD.DataController(Uplink, TargetJID);
            ActiveSession = NewSession;
            return true;
        }

        public async Task<LWTSD.DataController.DataPage> ReadData(int Page = 0)
        {
            if (RequestedResources.Count() == 0)
                throw new InvalidOperationException("No requested resources");

            if (KnownSchema != null)
            {
                foreach (var rkey in RequestedResources.Keys)
                {
                    if (!KnownSchema.Resources.ContainsKey(rkey))
                    {
                        KnownSchema = null;
                        ActiveSession = null;
                        break;
                    }

                }
            }

            if (ActiveSession == null || ActiveSession.SessionEndsAtUTC < DateTime.UtcNow.AddSeconds(15) )
            {
                await RequestNewSession(new TimeSpan(1, 0, 0));
            }



            return await LowlevelDataController.ReadData(KnownSchema,
                                                         RequestedResources.Keys.ToList(),
                                                         ActiveSession.ID);

        }

        public async Task<LWTSD.DataController.DataPage> ReadData(List<LWTSD.ResourcePath> Resources, int Page = 0)
        {
            if (RequestedResources.Count() == 0)
                throw new InvalidOperationException("No requested resources");

            foreach (var Res in Resources)
            {
                if (!RequestedResources.ContainsKey(Res))
                    throw new InvalidOperationException("Cannot request resource not in requested resources");
            }

            if (KnownSchema != null)
            {
                foreach (var rkey in RequestedResources.Keys)
                {
                    if (!KnownSchema.Resources.ContainsKey(rkey))
                    {
                        KnownSchema = null;
                        ActiveSession = null;
                        break;
                    }
                }
            }

            if (ActiveSession == null || ActiveSession.SessionEndsAtUTC < DateTime.UtcNow.AddSeconds(15))
            {
                await RequestNewSession(new TimeSpan(1, 0, 0));
            }


            return await LowlevelDataController.ReadData(KnownSchema,
                                                         RequestedResources.Keys.ToList(),
                                                         ActiveSession.ID,
                                                         Page);
        }

        public void Dispose()
        {
            SessionController.Dispose();
            LowlevelDataController.Dispose();
            Uplink = null;
        }


        public CDODataController(Connection Uplink,
                                 JID Orchestrator,
                                 string TargetEntityID,
                                 List<CDO.ResourceAccess> RequestedResources = null)
        {
            SessionController = new CDO.CDOController(Uplink, Orchestrator, null, SessionTerminated);
            this.TargetEntityID = TargetEntityID;
            this.Orchestrator = Orchestrator;
            this.Uplink = Uplink;

            if (RequestedResources != null)
                foreach (var res in RequestedResources)
                    this.RequestedResources[res.ResourcePath] = res;
        }
    }
}

File: typescript/CDOLWTSD/CDODataController.ts
import * as CDO from '../CDO/index';
import * as LWTSD from '../LWTSD/index';
import {AggregatedPresence, PresenceSubscribeResponse} from '../Misc/AggregatedPresence'

var xmpp = require('../node-xmpp/packages/client').xmpp;
var xml = require('../node-xmpp/packages/xml');

export class CDODataController
{
	SessionController : CDO.CDOController;
	LowlevelDataController : LWTSD.DataController | null;
	PresenceHandler : AggregatedPresence;
	Uplink : any;
	Orchestrator : string;
	TargetEntityID : string;

	public RequestedResources : Map<string, CDO.ResourceAccess> = new Map<string, CDO.ResourceAccess> ();

	KnownSchema : LWTSD.SimplifiedSchema | null;
	ActiveSession : CDO.Session | null;


	SessionTerminated(Session : CDO.Session) : void
	{
		if(!this.ActiveSession)
			return;

		if (Session.ID == this.ActiveSession.ID)
			this.ActiveSession = null;

		// TODO: check if any ongoing or pending new sessions towards JID ; if not - unsubscribe
	}

	public async RequestNewSession() 
	: Promise<boolean>
	{
		let NewSession = await this.SessionController.RequestSession(this.TargetEntityID, 
			Array.from(this.RequestedResources.values()), 
			60);
		if (NewSession == null)
			throw new Error("New session was actively denied");

		let BareJID = await this.SessionController.StartSession(NewSession.ID);

		let FullJids : Set<string> = await this.PresenceHandler.GetFullJIDs(BareJID);
		if(!FullJids || FullJids.size == 0)
		{
			throw new Error("Session failed to start; no full jid available");
		}

		let TargetJID = Array.from(FullJids.keys())[0];
		console.log("TargetJID = " + TargetJID);

		if (!TargetJID)
			throw new Error("Session failed to start");
		
		if (this.LowlevelDataController)
		{
			// TODO: unbind stuff against Client -- Dispose stuff...
			this.LowlevelDataController = null;
		}

		this.LowlevelDataController = new LWTSD.DataController(this.Uplink, TargetJID);
		this.ActiveSession = NewSession;
		return true;
	}

	public async ReadData(Page? : number, Resources? : Array<string>)
	: Promise<LWTSD.DataPage>
	{
		if(!Page)
			Page = 0;

		if (this.RequestedResources.size == 0)
			throw new Error("No requested resources");

		if (this.KnownSchema)
		{
			for (let rkey of Array.from(this.RequestedResources.keys()) )
			{
				if (!this.KnownSchema.Resources.has(rkey))
				{
					this.KnownSchema = null;
					this.ActiveSession = null;
					break;
				}

			}
		}

		if (this.ActiveSession == null)
		{
			let t = await this.RequestNewSession();
			// not use t..			
		}

		if(!Resources)
			Resources = Array.from(this.RequestedResources.keys());

		for (let Res of Resources)
		{
			if (!this.RequestedResources.has(Res))
				throw new Error("Cannot request resource not in requested resources");
		}

		if(!this.LowlevelDataController)
			throw new Error("LowlevelDataController is null")
		if(!this.ActiveSession)
			throw new Error("ActiveSession is null")

		/*if(!this.KnownSchema)
		{
			this.KnownSchema = await this.LowlevelDataController.GetSchema();
		}*/

		return await this.LowlevelDataController.ReadData(this.KnownSchema,
			Resources,
			this.ActiveSession.ID);

	}

	private HandlePresenceSubscribeRequest( From : string) : PresenceSubscribeResponse
	{
		return PresenceSubscribeResponse.AllowAndSubscribe;
	}
	
	public constructor(Uplink :any, 
		Orchestrator : string, 
		TargetEntityID : string, 
		RequestedResources? : Array<CDO.ResourceAccess> )
	{
		this.PresenceHandler = new AggregatedPresence(Uplink, this.HandlePresenceSubscribeRequest.bind(this));
		this.SessionController = new CDO.CDOController(Uplink, Orchestrator, this.SessionTerminated.bind(this));
		this.TargetEntityID = TargetEntityID;
		this.Orchestrator = Orchestrator;
		this.Uplink = Uplink;

		if (RequestedResources != null)
			for (let res of RequestedResources)
				this.RequestedResources.set(res.ResourcePath, res);
	}
}
